core/src/test/scala/io/delta/tables/DeltaTableBuilderSuite.scala

```diff
@@ -366,4 +375,101 @@ class DeltaTableBuilderSuite extends QueryTest with SharedSparkSession with Delt
         e.getMessage.contains("is not a valid name"))
     }
   }
+
+  test("delta table property case") {
+
+    val preservedCaseConfig = Map("delta.appendOnly" -> "true", "Foo" -> "Bar", "foo" -> "Bar")
+
+    val lowerCaseEnforcedConfig = preservedCaseConfig.map{
+      case (key, value) if !key.contains("delta") =>
+         /* Required since when reading from the snapshot the table property
+            will be merged with default configs using a case-insensitive table
+            and the value appendOnly from the default will dominate
+          */
+
+        (key.toLowerCase, value) // scalastyle:off caselocale
+      case x => x
+    }
+
+    sealed trait DeltaTablePropertySetOperation {
+
+      def setTableProperty(tablePath: String): Unit
+
+      def expectedConfig: Map[String, String]
+
+    }
+
+    trait CasePreservingTablePropertySetOperation extends DeltaTablePropertySetOperation {
+
+      val expectedConfig = preservedCaseConfig
+
+    }
+    case object SetPropertyThroughCreate extends CasePreservingTablePropertySetOperation {
+
+      def setTableProperty(tablePath: String): Unit = sql(
+        s"CREATE TABLE delta.`$tablePath`(id INT) " +
+          s"USING delta TBLPROPERTIES('delta.appendOnly'='true', 'Foo'='Bar', 'foo'='Bar' ) "
+      )
+
+    }
+    case object SetPropertyThroughAlter extends CasePreservingTablePropertySetOperation {
+
+      def setTableProperty(tablePath: String): Unit = {
+        spark.range(1, 10).write.format("delta").save(tablePath)
+        sql(s"ALTER TABLE delta.`$tablePath` " +
+          s"SET TBLPROPERTIES('delta.appendOnly'='true', 'Foo'='Bar', 'foo'='Bar')")
+      }
+
+    }
+
+    case class SetPropertyThroughTableBuilder(backwardCompatible: Boolean) extends
+      DeltaTablePropertySetOperation {
+
+      def setTableProperty(tablePath: String): Unit = {
+        val confItem = DeltaSQLConf.TABLE_BUILDER_FORCE_TABLEPROPERTY_LOWERCASE
+        val previousValue = spark.sessionState.conf.getConf(confItem)
+        try {
+          spark.conf.set(confItem.key, backwardCompatible)
+          DeltaTable.create()
+            .location(tablePath)
+            .property("delta.appendOnly", "true")
+            .property("Foo", "Bar")
+            .property("foo", "Bar")
+            .execute()
+        }
+        finally {
+          spark.conf.set(confItem.key, previousValue)
+        }
+
+      }
+
+      override lazy val expectedConfig : Map[String, String] = {
+        if (backwardCompatible) {
+          lowerCaseEnforcedConfig
+        }
+        else {
+          preservedCaseConfig
+        }
+      }
+    }
+    val examples = Seq(
+      SetPropertyThroughCreate,
+      SetPropertyThroughAlter,
+      SetPropertyThroughTableBuilder(true),
+      SetPropertyThroughTableBuilder(false)
+    )
+
+    for (example <- examples) {
```

Comment:
since these are essentially nested 'tests' within this one test case, can you use `withClue(...)` for each example, perhaps `withClue(example.description) { ... }`. this will make it easier to debug in the future, if we ever need to.

Reply:
💯